{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import fastas from Benchmark Genes folder\n",
    "import os\n",
    "import sys\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# get the path to the Benchmark Genes folder\n",
    "path = \"/Users/rishabjain/Desktop/Research/choformer/benchmarking/Benchmark Genes\"\n",
    "\n",
    "# get the list of all the files in the Benchmark Genes folder\n",
    "files = os.listdir(path)\n",
    "\n",
    "\n",
    "# get the list of all the fasta files in the Benchmark Genes folder\n",
    "fastas = [file for file in files if file.endswith(\".fasta\")]\n",
    "\n",
    "# get all the genes from each file (some files can have multiple)\n",
    "\n",
    "genes = []\n",
    "for fasta in fastas:\n",
    "    with open(path + \"/\" + fasta, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            genes.append(record)\n",
    "\n",
    "# get the list of all the genes\n",
    "gene_list = [gene.id for gene in genes]\n",
    "\n",
    "# get the list of all the gene sequences\n",
    "gene_sequences = [str(gene.seq) for gene in genes]\n",
    "\n",
    "# get the list of all the gene lengths\n",
    "gene_lengths = [len(gene) for gene in gene_sequences]\n",
    "\n",
    "# get the list of all the gene descriptions\n",
    "gene_descriptions = [gene.description for gene in genes]\n",
    "\n",
    "# write a final fasta file with all the genes\n",
    "with open(\"all_genes.fasta\", \"w\") as handle:\n",
    "    for gene in genes:\n",
    "        SeqIO.write(gene, handle, \"fasta\")\n",
    "\n",
    "# write a final text file with all the gene descriptions\n",
    "with open(\"all_gene_descriptions.txt\", \"w\") as handle:\n",
    "    for gene in gene_descriptions:\n",
    "        handle.write(gene + \"\\n\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'X'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 48\u001b[0m\n\u001b[1;32m     45\u001b[0m         proteins\u001b[38;5;241m.\u001b[39mappend(record)\n\u001b[1;32m     47\u001b[0m \u001b[38;5;66;03m# Reverse translate each gene sequence\u001b[39;00m\n\u001b[0;32m---> 48\u001b[0m gene_sequences_dna \u001b[38;5;241m=\u001b[39m [reverse_translate(gene) \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m proteins]\n\u001b[1;32m     50\u001b[0m \u001b[38;5;66;03m# Write a final FASTA file with all the genes\u001b[39;00m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall_genes_dna.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m handle:\n",
      "Cell \u001b[0;32mIn[7], line 48\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     45\u001b[0m         proteins\u001b[38;5;241m.\u001b[39mappend(record)\n\u001b[1;32m     47\u001b[0m \u001b[38;5;66;03m# Reverse translate each gene sequence\u001b[39;00m\n\u001b[0;32m---> 48\u001b[0m gene_sequences_dna \u001b[38;5;241m=\u001b[39m [\u001b[43mreverse_translate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m proteins]\n\u001b[1;32m     50\u001b[0m \u001b[38;5;66;03m# Write a final FASTA file with all the genes\u001b[39;00m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall_genes_dna.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m handle:\n",
      "Cell \u001b[0;32mIn[7], line 35\u001b[0m, in \u001b[0;36mreverse_translate\u001b[0;34m(protein_sequence)\u001b[0m\n\u001b[1;32m     33\u001b[0m dna_sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m amino_acid \u001b[38;5;129;01min\u001b[39;00m protein_sequence:\n\u001b[0;32m---> 35\u001b[0m     codons, probs \u001b[38;5;241m=\u001b[39m \u001b[43mfrequency\u001b[49m\u001b[43m[\u001b[49m\u001b[43mamino_acid\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     36\u001b[0m     \u001b[38;5;66;03m# Normalize the probabilities to sum to 1\u001b[39;00m\n\u001b[1;32m     37\u001b[0m     probs \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(probs) \u001b[38;5;241m/\u001b[39m np\u001b[38;5;241m.\u001b[39msum(probs)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'X'"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Codon usage table\n",
    "frequency = {\n",
    "    \"A\": ([\"GCG\", \"GCA\", \"GCT\", \"GCC\"], [0.11, 0.23, 0.26, 0.40]),\n",
    "    \"R\": ([\"AGG\", \"AGA\", \"CGG\", \"CGA\", \"CGT\", \"CGC\"], [0.20, 0.20, 0.21, 0.11, 0.08, 0.19]),\n",
    "    \"N\": ([\"AAT\", \"AAC\"], [0.46, 0.54]),\n",
    "    \"D\": ([\"GAT\", \"GAC\"], [0.46, 0.54]),\n",
    "    \"C\": ([\"TGT\", \"TGC\"], [0.45, 0.55]),\n",
    "    \"*\": ([\"TGA\", \"TAG\", \"TAA\"], [0.52, 0.20, 0.28]),\n",
    "    \"Q\": ([\"CAG\", \"CAA\"], [0.75, 0.25]),\n",
    "    \"E\": ([\"GAG\", \"GAA\"], [0.58, 0.42]),\n",
    "    \"G\": ([\"GGG\", \"GGA\", \"GGT\", \"GGC\"], [0.25, 0.25, 0.16, 0.34]),\n",
    "    \"H\": ([\"CAT\", \"CAC\"], [0.41, 0.59]),\n",
    "    \"I\": ([\"ATA\", \"ATT\", \"ATC\"], [0.16, 0.36, 0.48]),\n",
    "    \"L\": ([\"TTG\", \"TTA\", \"CTG\", \"CTA\", \"CTT\", \"CTC\"], [0.13, 0.07, 0.41, 0.07, 0.13, 0.20]),\n",
    "    \"K\": ([\"AAG\", \"AAA\"], [0.58, 0.42]),\n",
    "    \"M\": ([\"ATG\"], [1.0]),\n",
    "    \"F\": ([\"TTT\", \"TTC\"], [0.45, 0.55]),\n",
    "    \"P\": ([\"CCG\", \"CCA\", \"CCT\", \"CCC\"], [0.11, 0.27, 0.28, 0.33]),\n",
    "    \"S\": ([\"AGT\", \"AGC\", \"TCG\", \"TCA\", \"TCT\", \"TCC\"], [0.15, 0.24, 0.06, 0.15, 0.18, 0.22]),\n",
    "    \"T\": ([\"ACG\", \"ACA\", \"ACT\", \"ACC\"], [0.12, 0.28, 0.24, 0.36]),\n",
    "    \"W\": ([\"TGG\"], [1.0]),\n",
    "    \"Y\": ([\"TAT\", \"TAC\"], [0.43, 0.57]),\n",
    "    \"V\": ([\"GTG\", \"GTA\", \"GTT\", \"GTC\"], [0.47, 0.11, 0.18, 0.24])\n",
    "}\n",
    "\n",
    "# Reverse translate protein sequence to DNA sequence\n",
    "def reverse_translate(protein_sequence):\n",
    "    dna_sequence = \"\"\n",
    "    for amino_acid in protein_sequence:\n",
    "        codons, probs = frequency[amino_acid]\n",
    "        # Normalize the probabilities to sum to 1\n",
    "        probs = np.array(probs) / np.sum(probs)\n",
    "        dna_sequence += codons[np.random.choice(len(codons), p=probs)]\n",
    "    return dna_sequence\n",
    "\n",
    "# Read the proteins from all_genes.fasta\n",
    "proteins = []\n",
    "with open(\"all_genes.fasta\", \"r\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        proteins.append(record)\n",
    "\n",
    "# Reverse translate each gene sequence\n",
    "gene_sequences_dna = [reverse_translate(gene) for gene in proteins]\n",
    "\n",
    "# Write a final FASTA file with all the genes\n",
    "with open(\"all_genes_dna.fasta\", \"w\") as handle:\n",
    "    for i in range(len(gene_sequences_dna)):\n",
    "        gene = SeqRecord(Seq(gene_sequences_dna[i]), id=f\"gene_{i+1}\", description=f\"gene_{i+1}\")\n",
    "        SeqIO.write(gene, handle, \"fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
